design/deployment-strategies.md

```diff
@@ -0,0 +1,212 @@
+
+**routeTraffic** – Use to run tasks that serves the traffic to the updated version. The results of a lifecycle hook event can trigger a rollback.
+
+**postRouteTrafficChecK** - Use to run the tasks after the traffic is routed. Typically these tasks monitor the health of the updated version for defined interval. The results of a lifecycle hook event can trigger a rollback.
+
+**on: Failure or Success** - Use to run the task to peform rollback actions or clean-up. The results of a lifecycle hook event doesn't trigger a rollback.
+
+Using the lifecycle hooks, we can achieve complex deployment events such as - 
+
+- Blue Green
+- Canary
+- Rolling
+
+**Blue-Green**: Reduce deployment downtime by having identical standby environment. 
+At any time one of them, let's say `blue` for the example, is live. As you prepare a new release of your software,
+you do your final stage of testing in the green environment. Once the software is working in the green environment, 
+you switch the traffic so that all incoming requests go to the green environment - the blue one is now idle.
+
+Consider the following Azure Pipelines pipeline defined in YAML
+
+```yaml
+jobs:
+- deployment:
+  environment: musicCarnivalProd
+  pool:
+    name: musicCarnivalProdPool
+  strategy:                   
+    blueGreen:    
+      preDeploy:
+        steps:
+        - script: echo initialize, cleanup, install certs...
+      deploy:              
+        steps:                                   
+        - script: echo deploy app updates... 
+      routeTraffic:
+        delay: 60m
+        steps:
+        - script: echo route traffic...   
+      postRouteTafficCheck:
+        timeout: 60m
+        samplingInterval: 5m
+        steps:          
+        - script: echo monitor app health        
+      on:
+        failure:
+          steps:
+          - script: echo rollback, cleanup..     
+        success:
+          steps:
+          - script: echo checks passed...
+```
+
+For example, deploy a web app using blue-green strategy.
+
+```yaml
+jobs:
+- deployment:
+  environment: musicCarnivalProd.musicCarnivalWeb
+  pool:
+    name: musicCarnivalProdPool
+  strategy:                   
+    blueGreen:    
+      deploy:              
+        steps:                                   
+        - task: AzureWebApp@1
+          inputs:
+            appName: 'musicCarnivalWeb'
+            package: '$(System.DefaultWorkingDirectory)/**/*.zip' 
+            slotName: $(environment.staging)                        # deploy to green
+      routeTraffic:
+        delay: 60m
+        steps:
+        - task: AzureAppServiceManage@0
+          inputs:
+            Action: 'Swap Slots'                                    # swap green with blue
+            WebAppName: 'musicCarnivalWeb'
+            sourceSlot: $(environment.staging)
+            destinationSlot: $(environment.prod)
+      postRouteTafficCheck:
```

Comment:
Can we just call this postRouteTaffic instead?

Reply:
Good point, user can run any step after routing traffic not just checks. The pool type would determine, if it's agent or server or container